File: frontend/src/app/shared/models/term.model.ts
import { CourseModel } from './course.model';

export class TermModel {
  id: number | undefined;
  termName: string | undefined;
  active: boolean = false;
  termCourses: TermCourseModel[] | undefined;
  startDate: Date | undefined;
  endDate: Date | undefined;

  constructor(termModel: { [key: string]: any } | undefined) {
    this.id = termModel?.id;
    this.termName = termModel?.term_name;
    this.active = termModel?.active;
    this.termCourses = termModel?.term_courses?.map(
      (course: { [key: string]: any } | undefined) =>
        new TermCourseModel({ ...course })
    );
    this.startDate = new Date(termModel?.term_start);
    this.endDate = new Date(termModel?.term_end);
  }
}

export class TermCourseModel {
  id: number | undefined;
  term: TermModel | undefined;
  course: CourseModel | undefined;

  constructor(termCourseModel: { [key: string]: any } | undefined) {
    this.id = termCourseModel?.id;
    this.term = new TermModel({ ...termCourseModel?.term });
    this.course = new CourseModel({ ...termCourseModel?.course });
  }
}

File: backend/models/TermModel.py
from .db import db
from .CourseModel import Course


class TermCourse(db.Model):
    __tablename__ = 'term_course'
    id = db.Column(db.Integer, primary_key=True)
    course_id = db.Column(db.Integer, db.ForeignKey('course.id'), nullable=False)
    term_id = db.Column(db.Integer, db.ForeignKey('term.id'), nullable=False)
    course = db.relationship('Course', backref='term_courses')


class Term(db.Model):
    __tablename__ = 'term'
    id = db.Column(db.Integer, primary_key=True)
    term_name = db.Column(db.String(80), nullable=False)
    active = db.Column(db.Boolean, default=True)
    term_courses = db.relationship('TermCourse', backref='term')
    term_start = db.Column(db.DateTime)
    first_day_class = db.Column(db.DateTime)
    term_end = db.Column(db.DateTime)